### ISSUE TITLE
OpenSMTPD fails to start sometimes when "listen on all" is specified

### ISSUE THREAD

**Issue by wgraeber** [OpenSMTPD/OpenSMTPD#853]
I have a fairly simple OpenSMTPD configuration on OpenBSD 6.3 replicated across 3 servers. The listen portion is as follows:

listen on all tls pki mail.server.com auth-optional
listen on all port 587 tls-require pki mail.server.com auth senders <authorized>

I am running the servers on Vultr with a dual stack configuration, and /etc/hostname.vio0 looks like this:

dhcp
inet6 autoconf -autoconfprivacy

I recently notice on the "faster" server (dual core / 4 GB RAM) vs the two "slower" servers (single core, 1 GB RAM), that OpenSMTPD was failing to start. The error thrown was:

May  8 21:16:26 mail smtpd[16934]: info: OpenSMTPD 6.0.4 starting
May  8 21:16:27 mail smtpd[57207]: pony express: smtpd: bind: Can't assign requested address
May  8 21:16:27 mail smtpd[90309]: warn: lost child: pony express exited abnormally
May  8 21:16:27 mail smtpd[90309]: smtpd: process control socket closed

I believe this is probably somehow related to slaacd being slow to assign an ipv6 address. I was able to reproduce the error reliably when moving to ipv6 autoconf.

I found a workaround (it has at least persisted over several reboots now) was to change smtpd.conf as such:

listen on 0.0.0.0 tls pki mail.server.com auth-optional
listen on :: tls pki mail.server.com auth-optional
listen on 0.0.0.0 port 587 tls-require pki mail.server.com auth senders <authorized>
listen on :: tls pki mail.server.com auth-optional

**Comment by poolpOrg** [comment]
This can't be fixed because it is not really an issue with OpenSMTPD but rather with the networking layer and your workaround is really the best solution to avoid this.

The problem is that the `all` in `listen on all` is not a keyword, it is an interface group, similar to `listen on egress` or `listen on wlan`. That interface group doesn't equate to 0.0.0.0 or :: but it will contain all the IP addresses configured on attached interfaces when OpenSMTPD configures the daemon, new IP addresses will not be caught, and if there are things changing the IP addresses while the daemon is being configured, there's nothing we can do.

Your workaround is the real fix in your own setup.

**Comment by bob1de** [comment]
Hi,

I found this issue after I encountered the same problem on a Debian 11 host. The proposed workaround resolves it, however I was wondering whether config syntax could be extended to allow specifying a list of addresses/interfaces/interface groups to listen on, maybe like so:

    listen on { 0.0.0.0, :: } ...

At least this would make configurations less error-prone, in case one changes one of the duplicated listen statements but forgets to adapt the other.
Just to be clear, I don't expect it to support dynamic tables for the addresses, a static list would be sufficiently flexible IMHO.

I have no idea how hard that would be to implement, just thinking of a way to improve the already brilliant config syntax even more.

Thanks for considering.

Best regards
Robert
